### PR TITLE
Allow OAuth 2.0 user defined file sources (w/Dropbox integration)

### PR DESCRIPTION
**Pull request by jmchilton**
Builds on #18222 (only the last commit is relevant to this PR beyond that one currently).

This implements oauth2 client capabilities with the user defined file sources framework. There would be some work to integrate this with user defined object stores also but that would be pretty minimal given the code overlap/shared abstractions used on the client and server. 

This works currently more or less but I want to make sure it works with Scope requesting clients and on a second service so I going to wait until I've integrated a Google Drive plugin and documented that before taking it out of Draft/WIP.

This work integrates an initial oauth2 enabled plugin for Dropbox including admin documentation for how to configure it and how build and configure a Dropbox app within Dropbox for a Galaxy instance.

------
Documentation Screenshot 

Screenshot 2024-05-30 at 12 48 35 PM

--------
Documentation Screenshot 
Screenshot 2024-05-30 at 12 48 56 PM

--------
Documentation Screenshot 

Screenshot 2024-05-30 at 12 49 08 PM

--------

For oauth2 services - the plugin status framework has been updated to reflect some additional information. Hopefully we can use it down the road to know when refresh tokens need to be re-requested and such though nothing beyond initial creation and typical use is implemented here so far:

Screenshot 2024-05-30 at 12 53 18 PM

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  1. Follow the instructions in the admin docs to create a service and try it out. The docs say the oauth2 callback URI needs to be https but it doesn't need to be if the hostname is localhost. I use ``http://localhost:8081/oauth2_callback`` in my local testing with the ``make client-dev-server`` proxy.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by nuwang** [comment]
This is very cool! Let me know when it's ready and I'd be happy to do some testing.

File: lib/galaxy/webapps/galaxy/api/oauth2_callback.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    def oauth2_callback(
+        self,
+        trans: SessionRequestContext = DependsOnTrans,
+        state: str = StateQueryParam,
+        code: Optional[str] = CodeQueryParam,
+        error: Optional[str] = ErrorQueryParam,
+    ):
+        if error:
+            error_code = self._ensure_valid_oauth_error_code(error)
+            return RedirectResponse(f"{trans.request.url_path}{ERROR_REDIRECT_PATH}?error={error_code}")
```
[the remark this answers is not in the document]
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/api/oauth2_callback.py:62]: This seems like a false positive or am I missing something? `error_code` can only match the valid values now... :thinking:

**Review by itisAliRH — APPROVED** [review]
Thank you @jmchilton, this looks great. I appreciate it! 
I tested it, and it’s running perfectly. I made a few minor tweaks and will build on this to enhance the UI in future PRs.

**Review by davelopez — APPROVED** [review]
The remaining test failures are unrelated.

Thank you @jmchilton and @itisAliRH!

**Comment by jmchilton** [comment]
Wow - this is huge - thank you for the merge!

**Pull request merged**